File: lib/tokenizer/ClipTokenizer.dart
import 'ClipByteEncoder.dart';

class ClipTokenizer {
  final Map<String, int> encoder;
  final Map<StringPair, int> bpeRanks;

  ClipTokenizer(this.encoder, this.bpeRanks);

  static final RegExp encodeRegex = RegExp(
      r"\[PAD\]|\[CLS\]|\[SEP\]|\[MASK\]|[A-Za-z]+|\[unused\d+\]|[^\sA-Za-z]+",
      unicode: true);

  List<int> encode(String text) {
  // Tokenize text using larger tokens from vocab.json
  return encodeRegex.allMatches(text).expand((match) {
    final token = match.group(0)!;
    if (encoder.containsKey(token)) {
      return [encoder[token]!];
    }
    // Apply byte encoding and BPE if token is not directly found
    final byteEncoded = byteEncode(token);
    return bpe(byteEncoded)
        .map((t) => encoder[t] ?? encoder['[UNK]'] ?? -1) // Ensure no null values
        .where((value) => value != -1) // Filter out fallback if it remains unset
        .toList();
  }).toList();
}


  String byteEncode(String text) {
    // Convert characters to byte-encoded string
    return String.fromCharCodes(text.runes.map((codePoint) => byteEncoder[codePoint]!.codeUnitAt(0)));
  }

  List<String> bpe(String token) {
    if (token.length <= 1 || encoder.containsKey(token)) return [token];
    var word = token.split('').toList();
    var pairs = getPairs(word);
    while (pairs.isNotEmpty) {
      final validPairs = pairs.where((pair) => bpeRanks.containsKey(pair)).toList();
      if (validPairs.isEmpty) break;
      final pair = validPairs.reduce((a, b) => bpeRanks[a]! < bpeRanks[b]! ? a : b);
      word = mergePairInWord(word, pair.item1, pair.item2);
      pairs = getPairs(word);
    }
    return word;
  }

  List<String> mergePairInWord(List<String> word, String first, String second) {
    var newWord = <String>[];
    int i = 0;
    while (i < word.length) {
      final j = word.indexWhere((token) => token == first, i);
      if (j != -1) {
        newWord.addAll(word.sublist(i, j));
        if (j < word.length - 1 && word[j + 1] == second) {
          newWord.add(first + second);
          i = j + 2;
        } else {
          newWord.add(word[j]);
          i = j + 1;
        }
      } else {
        newWord.addAll(word.sublist(i));
        break;
      }
    }
    return newWord;
  }

  Set<StringPair> getPairs(List<String> word) {
    final pairs = <StringPair>{};
    for (int i = 0; i < word.length - 1; i++) {
      pairs.add(StringPair(word[i], word[i + 1]));
    }
    return pairs;
  }
}

class StringPair {
  final String item1;
  final String item2;

  StringPair(this.item1, this.item2);

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is StringPair && runtimeType == other.runtimeType &&
      item1 == other.item1 && item2 == other.item2;

  @override
  int get hashCode => item1.hashCode ^ item2.hashCode;
}

File: android/app/src/main/kotlin/com/example/photos/tokenizer/ClipTokenizer.kt
//package com.example.photos.tokenizer
//
//
//import android.os.Build
//import androidx.annotation.RequiresApi
//import java.util.regex.Pattern
//
//class ClipTokenizer(
//    private val encoder: Map<String, Int>,
//    private val bpeRanks: Map<Pair<String, String>, Int>
//) {
//    companion object {
//        private val encodeRegex = Pattern.compile(
//            "\\[PAD\\]|\\[CLS\\]|\\[SEP\\]|\\[MASK\\]|[A-Za-z]+|\\[unused\\d+\\]|[^\\sA-Za-z]+",
//            Pattern.UNICODE_CHARACTER_CLASS
//        )
//    }
//
//    @RequiresApi(Build.VERSION_CODES.UPSIDE_DOWN_CAKE)
//    fun encode(text: String): List<Int> {
//        // Tokenize text using larger tokens from vocab.json
//        return encodeRegex.matcher(text).results()
//            .mapNotNull { match ->
//                val token = match.group()
//                encoder[token]?.let { listOf(it) }
//                    ?: bpe(byteEncode(token))
//                        .mapNotNull { encoder[it] ?: encoder["[UNK]"] }
//            }
//            .toList()
//    }
//
//    private fun byteEncode(text: String): String {
//        // Convert characters to byte-encoded string
//        return text.codePoints()
//            .mapToObj { codePoint -> byteEncoder[codePoint] }
//            .joinToString("") { it ?: "" }
//    }
//
//    private fun bpe(token: String): List<String> {
//        if (token.length <= 1 || encoder.containsKey(token)) return listOf(token)
//        var word = token.split("").toMutableList()
//        var pairs = getPairs(word)
//        while (pairs.isNotEmpty()) {
//            val validPairs = pairs.filter { bpeRanks.containsKey(it) }
//            if (validPairs.isEmpty()) break
//            val pair = validPairs.minByOrNull { bpeRanks[it]!! }!!
//            word = mergePairInWord(word, pair.first, pair.second)
//            pairs = getPairs(word)
//        }
//        return word
//    }
//
//    private fun mergePairInWord(word: MutableList<String>, first: String, second: String): MutableList<String> {
//        val newWord = mutableListOf<String>()
//        var i = 0
//        while (i < word.size) {
//            val j = word.indexOfFirst(first, i)
//            if (j != -1) {
//                newWord.addAll(word.subList(i, j))
//                if (j < word.size - 1 && word[j + 1] == second) {
//                    newWord.add("$first$second")
//                    i = j + 2
//                } else {
//                    newWord.add(word[j])
//                    i = j + 1
//                }
//            } else {
//                newWord.addAll(word.subList(i, word.size))
//                break
//            }
//        }
//        return newWord
//    }
//
//    private fun getPairs(word: List<String>): Set<Pair<String, String>> {
//        return word.windowed(2, 1)
//            .map { it[0] to it[1] }
//            .toSet()
//    }
//}